optex/base/optex.lua

```diff
@@ -364,3 +374,242 @@ callback.add_to_callback("input_level_string",
+            local new = getattribute(n, color_attribute) or 0
+            if new ~= current then
+                head    = insertbefore(head, n, copy(colors[new]))
+                current = new
+            end
+        end
+    end
+    return head, current
+end
+--
+-- For \TeX/ end, we define \`\_setcolor``{<color literal>}`, which translates
+-- <color literal> to number using `colors` table (allocating a new number if
+-- necessary) and sets the attribute to the number.
+--
+-- If we are setting to `0` (most probably due to `\Black`), we instead unset
+-- the attribute entirely. Both would have been handled the same way by
+-- `colorize`, but this way we save on attribute nodes.
+--
+define_lua_command("_setcolor", function()
+    local color = token_scanargument()
+    local num = colors[color]
+    if not num then
+        num = #colors + 1
+        colors[num] = pdfliteral(color)
+        colors[color] = num
+    elseif num == 0 then
+        -- don't bother setting the attribute to zero, when unsetting it has the same effect
+        num = -0x7FFFFFFF
+    end
+    setattribute(color_attribute, num)
+end)
+--
+-- In similar fashion, we define \`\_resetcolor`, which unsets the attribute,
+-- hence restoring initial color (probably black).
+--
+define_lua_command("_resetcolor", function()
+    setattribute(color_attribute, -0x7FFFFFFF)
+end)
+--
+-- Because it is so easy to use different default/initial color other than
+-- black, we allow it. Of course we need to change what `colors` maps to/from
+-- `0`. Furthermore we need to note that black is special, because that is what
+-- all pages start on. For any other color we need to force an initial color setting in `colorize`,
+-- regardless of what color is currently set. This is handled by
+-- `initial_color`, the initial value of `colorize`'s `current`. For black it
+-- can be `0` (current/initial color is indeed black), but if set to something
+-- not equal to any number assigned to color (we use `-1`) initial color is set
+-- to color `0` (now most probably non-black).
+--
+local initial_color = 0
+define_lua_command("_setdefaultcolor", function()
+    local color = token_scanargument()
+    colors[0] = pdfliteral(color)
+    colors[color] = 0
+     -- not zero, so initial color will be set on every page on first content encountered
+    initial_color = -1
+end)
+--
+-- Now we need to somehow allow \TeX/ to use `colorize` on otherwise finished
+-- boxes. We do this by redefining `\_shipout`, which means that our mechanism
+-- works almost transparently to the end user. This is sadly not really possible
+-- with \pdfxform, which creates PDF form XObjects (very similiar to PDF page
+-- content). Redefining the said primitive is not sufficient, because of
+-- `\immediate`. Those that want to colorize boxes before transforming them to
+-- xforms should use \`\_colorizebox``<boxnum>`.
+--
+define_lua_command("_shipout", function()
```

Comment:
Custom callback instead?

Reply:
What about to keep the primitive `\shipout` meaning and create a pseudo-primitive `\boxprepare <num> <box>` which does colorization of given box and set ti to the box<num>. Then we can declare:
```
\_def \_optexoutput{\_begoutput \_boxprepare0\_completepage \_shipout\_box0 \_endoutput}
```
Of course, if somebody created his own \output routine then the colorization will be missing in such s routine. But I hope that today, there is no more individual output routines for OpTeX.
Advantage: somebody will want to add cop-makrs, for example:
```
\_def \_optexoutput{\_begoutput \_boxprepare0\_completepage \_setbox0=\_addcropmarks{\_box0}\_shipout\_box0 \_endoutput}
```

Reply:
I have these changes ready. Let me know when I can safely push.

I did it more generally and called it `\preshipout`. Using callback mechanism more things could be added, other than colors.

Current output routine with these changes is:

```
\_def \_optexoutput{\_begoutput \_preshipout0\_completepage \_shipout\_box0 \_endoutput}
```

And for example in `pdfextra` it looks like this:

```diff
-   \_colorizebox0
+   \_preshipout0\_box0
    \_immediate\_pdfxform0
```

I.e. we can remove `\_colorizebox` and instead use `\_preshipout` for both `\shipout` and `\pdfxform` (they execute same `ship_out` procedure anyways). By the way, we can add a keyword to make it more readable. Maybe something like `\_preshipout tobox0 \_completepage`?

But to be honest I would have prefered:

```
\_def \_optexoutput{\_begoutput \_setbox0=\_completepage \preshipout0 \_shipout\_box0 \_endoutput}
```

and

```diff
    \_preshipout0
    \_immediate\_pdfxform0
```